lib/Sema/TypeCheckPattern.cpp

```diff
@@ -736,6 +736,62 @@ static bool validateParameterType(ParamDecl *decl, DeclContext *DC,
     }
     decl->getTypeLoc().setType(Ty);
   }
+  // If the param is not a 'let' and it is not an 'inout'.
+  // It must be a 'var'. Provide helpful diagnostics like a shadow copy
+  // in the function body to fix the 'var' attribute
+  if (!decl->isLet() && !Ty->is<InOutType>()) {
+    auto innermostMethodCtx = dyn_cast_or_null<AbstractFunctionDecl>(DC);
+    if (!innermostMethodCtx || !innermostMethodCtx->hasBody()) {
+      // If there is not function body, just suggest removal of 'var'
+      TC.diagnose(decl->getLetVarInOutLoc(), diag::var_parameter_not_allowed)
+      .fixItRemove(decl->getLetVarInOutLoc());
+    } else {
+      // Insert the shadow copy. The computations that follow attempt to
+      // 'best guess' the indentation and new lines so that the user
+      // doesn't have to add any whitespace
+      auto declBody = innermostMethodCtx->getBody();
+      SourceLoc insertionStartLoc;
+      if (!declBody->getNumElements()) {
+        insertionStartLoc = declBody->getLBraceLoc();
+      } else {
+        insertionStartLoc = declBody->getElement(0).getStartLoc();
+      }
+      if (insertionStartLoc.isInvalid()) {
+        TC.diagnose(decl->getLetVarInOutLoc(), diag::var_parameter_not_allowed)
+        .fixItRemove(decl->getLetVarInOutLoc());
+      } else {
+        auto startingString = CharSourceRange(TC.Context.SourceMgr,
+                                              declBody->getLBraceLoc(),
+                                              insertionStartLoc).str().str();
+        bool sawNewline = false;
```

Comment:
I'm really new to compilers so I didn't realize that there was a column location on SourceMgr. But, in any case how else could we determine whether to insert spaces or tabs for the indentation with just the column location info?